### ISSUE TITLE
Support for TF2.x Eager Execution

### ISSUE THREAD

**Issue by seanpmorgan** [IBM/adversarial-robustness-toolbox#238]
**Is your feature request related to a problem? Please describe.**
Currently Keras models written in TF2.x will fail during wrapping of `KerasClassifier`. At a minimum ART should be able to test TF2.x Keras models that are able to `model.compile` since the calls to e.g. `K.gradients()` should still work. On the dev 1.1.0 branch there is no support for using `tf.keras.applications.model*`:
https://colab.research.google.com/drive/1W_BpUW6RilGErmDXm5B9OWhcXIkUOQjf

Looking forward though, I'd be interested in determining the path / working on supporting full eager models.

**Additional context**
Is there a roadmap for supporting this? Or would ART be interested in a design proposal to support it?

**Comment by beat-buesser** [comment]
Hi @seanpmorgan Thank you very much for opening this issue. Your feedback is very much appreciated and I'll try to provide an overview of the current status.

`KerasClassifier` does not really fail, it stops intentionally with an exception if it detects eager execution, because officially it does not support eager execution. Currently you have to deactivate eager execution for `KerasClassifier` and `tf.keras` for TF2.0.

When we tested `KerasClassifier` with the release candidates of TF2.0 we got errors around `K.gradients()` saying that it would not be available for eager execution. I would have to check if this has changed since the official release.

Therefore we have created `TensorFlowV2Classifier` which supports eager execution and uses all the new features of TF2.0. `TensorFlowV2Classifier` should be able to support `tf.keras.applications.model.*` in eager execution mode.

Currently we are looking for user feedback to see if changing/refactoring `KerasClassifier` classifier to support eager execution is something that would needed or if TensorFlowV2Classifier` is able to provide all the required functionalities for TF2.0.

**Comment by beat-buesser** [comment]
This is an example of using ART 1.x with TensorFlow2.x, eager execution, and tf.keras.applications.model.* (ResNet50), I'm planning to add it to `examples`:
```
#!/usr/bin/env python
# coding: utf-8

import tensorflow as tf
from tensorflow.keras.preprocessing import image
from tensorflow.keras.applications.resnet50 import preprocess_input, decode_predictions
import numpy as np
from matplotlib import pyplot as plt
import imagenet_stubs

from art.classifiers import TensorFlowV2Classifier
from art.attacks import FastGradientMethod

# Check if executing eagerly
print('Executing eagerly:', tf.executing_eagerly())

# Get an image of a tractor
image_name = 'tractor.jpg'
x_image = None

for image_path in imagenet_stubs.get_image_paths():
    im = image.load_img(image_path, target_size=(224, 224))
    im = image.img_to_array(im)
    im = np.expand_dims(im, axis=0)

    if image_path.endswith(image_name):
        x_image = im

plt.imshow(x_image[0] / 255)
plt.show()

# Get ResNet50 model from tf.keras
# ResNet50 should predict `tractor` with probability 0.80

model = tf.keras.applications.resnet50.ResNet50(weights='imagenet')

x = x_image[0].copy()
x = image.img_to_array(x)
x = np.expand_dims(x, axis=0)
x = preprocess_input(x)

preds = model(x).numpy()

print('Predicted by ResNet50 model:', decode_predictions(preds, top=3)[0])

# Create ART classifier for TensorFlow2.x and eager execution

loss_object = tf.keras.losses.SparseCategoricalCrossentropy()
nb_classes = 1000
# Define preprocessing using the mean values defined in from tensorflow.keras.applications.resnet50.preprocess_input
# Expecting input images in BGR format
preprocessing = (np.asarray([103.939, 116.779, 123.68]), 1)
clip_values = (0, 255)

# Create ART classifier
classifier = TensorFlowV2Classifier(model, nb_classes=nb_classes, loss_object=loss_object,
                                    preprocessing=preprocessing,
                                    clip_values=clip_values)

# Evaluate ART classifier
x = x_image[0].copy()
x = image.img_to_array(x)
x = np.expand_dims(x, axis=0)

# Covnert RGB to BGR
x = x[..., ::-1]

preds = classifier.predict(x)
print('Predicted by ART classifier:', decode_predictions(preds, top=3)[0])

# Create FGSM attack and create adversarial example for tractor
attack = FastGradientMethod(classifier=classifier, norm=np.inf, eps=8, eps_step=8, targeted=False,
                            num_random_init=0, batch_size=1, minimal=False)

x_adv = attack.generate(x)

# ResNet50 model now predicts `plow` with probability 0.39
preds_adv = model(x_adv).numpy()
print('Predicted by ResNet50 model on adversarial example:', decode_predictions(preds_adv, top=3)[0])

# Plot adversarial example
plt.imshow(x_adv[0][..., ::-1] / 255)

# Calculate L_inf of perturbation, should be equal to eps
l_inf = np.amax(x_adv[0, :, :, 0] - x[0, :, :, 0])
print('Perturbation l_inf norm: %d' % l_inf)
```

**Comment by seanpmorgan** [comment]
Thanks for the followup @beat-buesser. Closing this issue as eager execution is supported! I'm going to start another issue to discuss the Keras support into the future.
